File: php/tests/SubscriptionServerTest.php
<?php

declare(strict_types=1);

namespace SolanaMpp\Tests;

use DateTimeImmutable;
use PHPUnit\Framework\TestCase;
use SolanaMpp\Core\Challenge;
use SolanaMpp\Core\Credential;
use SolanaMpp\Core\Headers;
use SolanaMpp\Intent\SubscriptionRequest;
use SolanaMpp\Server\SubscriptionServer;
use SolanaMpp\Server\SubscriptionVerifier;
use SolanaMpp\Server\VerificationResult;

final class SubscriptionServerTest extends TestCase
{
    public function testCreatesChallengeHeaderAndVerifiesCredential(): void
    {
        $server = new SubscriptionServer(secretKey: 'secret', realm: 'api');
        $request = new SubscriptionRequest(
            amount: '1000',
            currency: 'USDC',
            periodUnit: SubscriptionRequest::PERIOD_MONTH,
            periodCount: '1',
            externalId: 'subscription-001',
        );
        $challenge = Headers::parseWwwAuthenticate($server->createChallengeHeader($request));
        $credential = new Credential(
            challenge: $challenge->toEcho(),
            payload: ['type' => 'subscription-activation', 'signature' => 'sig'],
        );

        $result = $server->verifyAuthorizationHeader(
            $credential->toAuthorizationHeader(),
            new class implements SubscriptionVerifier {
                public function verify(Credential $credential, Challenge $challenge): VerificationResult
                {
                    TestCase::assertSame('sig', $credential->payload['signature']);
                    TestCase::assertSame('subscription', $challenge->intent);

                    return VerificationResult::success(reference: 'subscription-id', externalId: 'subscription-001');
                }
            },
        );

        self::assertTrue($result->ok);
        self::assertSame('subscription-id', $result->reference);
        $receipt = Headers::parseReceipt($server->createReceiptHeader($challenge, $result));
        self::assertSame($challenge->id, $receipt->challengeId);
        self::assertSame('subscription-001', $receipt->externalId);
    }

    public function testRejectsWrongIntent(): void
    {
        $server = new SubscriptionServer(secretKey: 'secret', realm: 'api');
        $request = (new SubscriptionRequest(
            amount: '1000',
            currency: 'USDC',
            periodUnit: SubscriptionRequest::PERIOD_MONTH,
            periodCount: '1',
        ))->toArray();
        $encodedRequest = Challenge::withSecret(
            secretKey: 'secret',
            realm: 'api',
            method: 'solana',
            intent: 'subscription',
            request: $request,
        )->request;
        $challenge = new Challenge(
            id: Challenge::computeId(
                secretKey: 'secret',
                realm: 'api',
                method: 'solana',
                intent: 'charge',
                request: $encodedRequest,
            ),
            realm: 'api',
            method: 'solana',
            intent: 'charge',
            request: $encodedRequest,
        );
        $credential = new Credential(challenge: $challenge->toEcho(), payload: ['type' => 'subscription-activation']);

        $result = $server->verifyAuthorizationHeader(
            $credential->toAuthorizationHeader(),
            $this->unusedVerifier(),
        );

        self::assertFalse($result->ok);
        self::assertSame('challenge method or intent mismatch', $result->reason);
    }

    public function testRejectsExpiredChallenge(): void
    {
        $server = new SubscriptionServer(secretKey: 'secret', realm: 'api');
        $challenge = $server->createChallenge(
            new SubscriptionRequest(
                amount: '1000',
                currency: 'USDC',
                periodUnit: SubscriptionRequest::PERIOD_MONTH,
                periodCount: '1',
            ),
            expires: '2026-01-01T00:00:00+00:00',
        );
        $credential = new Credential(challenge: $challenge->toEcho(), payload: ['type' => 'subscription-activation']);

        $result = $server->verifyAuthorizationHeader(
            $credential->toAuthorizationHeader(),
            $this->unusedVerifier(),
            new DateTimeImmutable('2026-05-19T00:00:00+00:00'),
        );

        self::assertFalse($result->ok);
        self::assertSame('challenge expired', $result->reason);
    }

    private function unusedVerifier(): SubscriptionVerifier
    {
        return new class implements SubscriptionVerifier {
            public function verify(Credential $credential, Challenge $challenge): VerificationResult
            {
                TestCase::fail('verifier should not be called');
            }
        };
    }
}

File: ruby/test/subscription_server_test.rb
# frozen_string_literal: true

require_relative 'test_helper'

module SolanaMpp
  module Server
    class SubscriptionServerTest < Minitest::Test
      def test_creates_challenge_header_and_verifies_credential
        server = SubscriptionServer.new(secret_key: 'secret', realm: 'api')
        request = Intents::SubscriptionRequest.new(
          amount: '1000',
          currency: 'USDC',
          period_unit: Intents::SubscriptionRequest::PERIOD_MONTH,
          period_count: '1',
          external_id: 'subscription-001'
        )
        challenge = Core::Headers.parse_www_authenticate(server.create_challenge_header(request))
        credential = Core::Credential.new(
          challenge: challenge.to_echo,
          payload: { 'type' => 'subscription-activation', 'signature' => 'sig' }
        )

        result = server.verify_authorization_header(
          credential.to_authorization_header,
          verifier: lambda do |parsed_credential, parsed_challenge|
            assert_equal 'sig', parsed_credential.payload['signature']
            assert_equal 'subscription', parsed_challenge.intent
            VerificationResult.success(reference: 'subscription-id', external_id: 'subscription-001')
          end
        )

        assert result.ok?
        assert_equal 'subscription-id', result.reference
        receipt = Core::Headers.parse_receipt(server.create_receipt_header(challenge, result))
        assert_equal challenge.id, receipt.challenge_id
        assert_equal 'subscription-001', receipt.external_id
      end

      def test_rejects_wrong_intent
        server = SubscriptionServer.new(secret_key: 'secret', realm: 'api')
        request = Intents::SubscriptionRequest.new(
          amount: '1000',
          currency: 'USDC',
          period_unit: Intents::SubscriptionRequest::PERIOD_MONTH,
          period_count: '1'
        )
        charge_challenge = Core::Challenge.with_secret(
          secret_key: 'secret',
          realm: 'api',
          method: 'solana',
          intent: 'charge',
          request: request.to_h
        )
        credential = Core::Credential.new(challenge: charge_challenge.to_echo, payload: { 'type' => 'subscription-activation' })

        result = server.verify_authorization_header(
          credential.to_authorization_header,
          verifier: unused_verifier
        )

        refute result.ok?
        assert_equal 'challenge method or intent mismatch', result.reason
      end

      def test_rejects_expired_challenge
        server = SubscriptionServer.new(secret_key: 'secret', realm: 'api')
        challenge = server.create_challenge(
          Intents::SubscriptionRequest.new(
            amount: '1000',
            currency: 'USDC',
            period_unit: Intents::SubscriptionRequest::PERIOD_MONTH,
            period_count: '1'
          ),
          expires: '2026-01-01T00:00:00Z'
        )
        credential = Core::Credential.new(challenge: challenge.to_echo, payload: { 'type' => 'subscription-activation' })

        result = server.verify_authorization_header(
          credential.to_authorization_header,
          verifier: unused_verifier,
          now: Time.iso8601('2026-05-19T00:00:00Z')
        )

        refute result.ok?
        assert_equal 'challenge expired', result.reason
      end

      private

      def unused_verifier
        lambda do |_credential, _challenge|
          flunk 'verifier should not be called'
        end
      end
    end
  end
end
